src/client/resources/web/js/chat.mjs

```diff
@@ -122,6 +123,317 @@ const serverInfo = {
     }
 };
 
+/**
+ * Player list
+ */
+
+
+/**
+ * Extends PlayerInfo to include the cached player head image and DOM element.
+ * @typedef {PlayerInfo & {
+*   playerHead?: string | null, // Cached base64 image data for the player's head.
+*   element?: HTMLElement | null, // Cached reference to the player's DOM element.
+*   playerClickHandler?: EventListener // Reference to the click event handler for cleanup.
+* }} StoredPlayerInfo
+*/
+
+/**
+* Interface for the player list manager.
+* @typedef {Object} PlayerListManager
+* @property {Map<string, StoredPlayerInfo>} players - Map of player IDs to player data.
+* @property {(newPlayerList: PlayerInfo[]) => Promise<void>} updatePlayerList - Updates the player list.
+* @property {(player: StoredPlayerInfo, existingPlayer?: StoredPlayerInfo) => HTMLElement | null} updatePlayerElement - Updates a player's DOM element.
+* @property {(playerId: string) => void} removePlayerElement - Removes a player's DOM element.
+* @property {(playerId: string) => StoredPlayerInfo | null} getPlayer - Gets a player by ID.
+* @property {() => StoredPlayerInfo[]} getAllPlayers - Gets all players.
+* @property {() => number} getPlayerCount - Gets the total player count.
+* @property {() => void} clearAll - Removes all players and clears the DOM.
+*/
+
+/**
+* Creates a player list manager that handles both data storage and DOM updates.
+* @param {HTMLElement} listContainer - The container element where player elements will be rendered.
+* @returns {PlayerListManager} An object with methods to manage the player list.
+*/
+const createPlayerList = (listContainer) => {
+    /**
+     * Flag to indicate if an update is in progress if set to true updating will be skipped.
+     * This shouldn't be too big of an issue once the initial playerlist is created as updates will be send every few seconds.
+     */
+    let isUpdating = false;
+
+    return {
+        /** @type {Map<string, StoredPlayerInfo>} */
+        players: new Map(),
+
+        /**
+         * Updates the player list. Makes sure that player info is updated,
+         * new players are added, and removed players are cleaned up.
+         *
+         * @param {PlayerInfo[]} newPlayerList - The new list of players.
+         * @returns {Promise<void>} Resolves when the update process is complete.
+         */
+        async updatePlayerList(newPlayerList) {
```

Comment:
One nice thing about using classes is they give tools like private methods to encourage smaller functions. Sure this objects we have in chat.mjs work find and we don't need more than one player list instance at a time, but classes are simply more feature-rich than old-fashioned JS objects.

Reply:
It did occur to me when making this that maybe classes would be nicer. Not going to refactor that now though, something for the backlog ;)

Reply:
> for the backlog

Where ideas go to die :P

Reply:
> Where ideas go to die :P

So far 12 open 10 closed, doing pretty well there :P